bundler/lib/bundler/source/path/installer.rb

```diff
@@ -26,18 +26,16 @@ def initialize(spec, options = {})
         end
 
         def post_install
-          SharedHelpers.chdir(@gem_dir) do
```

Comment:
Removing this `chdir` slightly changes the semantics of the `:pre_install`, `:post_build`, and `:post_install` hooks -- if anyone else is using those hooks and expecting the current directory while running the hook to be `@gem_dir` this could break compatibility.

Are these hooks considered part of a public API, or are they only used internally by `bundler` itself?

Reply:
These are hooks originally run and registered by `rubygems` and the thing is, they don't actually switch to the installed gem folder when run by `rubygems` on a regular `gem install <gem_name>` command, or when run through `Gem::Installer` on other `bundler` sources, like the standard `source "https://rubygems.org`.

In this case, we run them manually because for the `git` source we don't use `Gem::Installer` to install the gem, but I don't see a reason why we should switch to the gem folder for this particular source.

Reply:
In order words, in my eyes this as a bug fix, not a breaking change.

If some gem `foo` has a gem hook registered that depends on being on the gem's installation directory, the hook wouldn't be working at all, because `gem install` does not switch folders before running gem hooks.

Reply:
Makes sense!

FWIW, in researching this warning message I ran across this:

https://bugs.ruby-lang.org/issues/15661

That issue points out that `Dir.chdir` has a _process-global_ effect. This has interesting implications for Bundler::Worker and parallel installs.

I'm sure that's been considered already, since I see some integration with the rubygems `CHDIR_MUTEX` - but nonetheless I plan to do a quick audit of the bundler code to make sure that all use of `chdir` is safe and won't have unexpected side effects if multiple workers are running in parallel.

Reply:
Yeah, globally changing the PWD for compiling extensions is the biggest thread safety concern we have. It should be considered as you point out, but I'm not sure if everything is fully thread safe since we have some flaky test failures that seem to indicate the opposite.

I started some work on rubygems in relation to that which consists of no longer changing folders inside the ruby process when compiling extensions, but use `make -C`, `ruby -C` and so on instead.

Reply:
Your audit will be super appreciated by the way :heart_eyes:.

Reply:
Another possible way to address it could be to fork before building native extensions - the child process could then chdir all it wants without affecting anything else.

Reply:
I guess, yeah.

So this PR looks good to you, right?

Reply:
Yep, looks good